state/upgrades.go

```diff
@@ -1939,3 +1939,55 @@ func EnsureDefaultModificationStatus(pool *StatePool) error {
 	}
 	return nil
 }
+
+// EnsureApplicationDeviceConstraints ensures that there is a device
+// constraints document for every application.
+func EnsureApplicationDeviceConstraints(pool *StatePool) error {
+	st := pool.SystemState()
+	db := st.db()
+
+	applicationCol, applicationCloser := db.GetRawCollection(applicationsC)
+	defer applicationCloser()
+	applicationIter := applicationCol.Find(nil).Iter()
+	defer applicationIter.Close()
+
+	constraintsCol, constraintsCloser := db.GetRawCollection(deviceConstraintsC)
+	defer constraintsCloser()
+
+	var ops []txn.Op
+	var application applicationDoc
+	for applicationIter.Next(&application) {
+		// Since we are using a raw collection, we need to manually
+		// ensure that we prefix the IDs with the model-uuid.
+		localID := applicationDeviceConstraintsKey(application.Name, application.CharmURL)
+		key := ensureModelUUID(application.ModelUUID, localID)
+
+		// We only need to migrate applications that don't have a device
+		// constraints document. So we need to first check if there is one, before
+		// creating a txn.Op for the missing document.
+		var doc statusDoc
+		err := constraintsCol.Find(bson.D{{"_id", key}}).Select(bson.D{{"_id", 1}}).One(&doc)
+		if err == nil {
+			continue
+		} else if err != mgo.ErrNotFound {
+			return errors.Trace(err)
+		}
+
+		rawDoc := deviceConstraintsDoc{
```

Comment:
What happens if you just insert an empty deviceConstraintsDoc? And inline would be fine. I'm fairly sure it'll work.

Reply:
ok